examples/Reporting/ParallelReportDownload.cs

```diff
@@ -0,0 +1,242 @@
+        private readonly string[] GAQL_QUERY_STRINGS =
+        {
+            @"SELECT campaign.id, metrics.impressions, metrics.clicks
+                FROM campaign
+                WHERE segments.date DURING LAST_30_DAYS",
+            @"SELECT campaign.id, ad_group.id, metrics.impressions, metrics.clicks
+                FROM ad_group
+                WHERE segments.date DURING LAST_30_DAYS"
+        };
+
+        /// <summary>
+        /// Main method, to run this code example as a standalone application.
+        /// </summary>
+        /// <param name="args">The command line arguments.</param>
+        public static void Main(string[] args)
+        {
+            ParallelReportDownload codeExample = new ParallelReportDownload();
+            Console.WriteLine(codeExample.Description);
+
+            // The Google Ads customer IDs for which the call is made.
+            string customerIds = "INSERT_CUSTOMER_ID_1_HERE, INSERT_CUSTOMER_ID_2_HERE";
+
+            // Optional login customer ID if your access to the CIDs is via a manager account.
+            string loginCustomerId = "INSERT_MANAGER_ACCOUNT_ID_HERE";
+
+            codeExample.Run(new GoogleAdsClient(), customerIds, loginCustomerId);
+        }
+
+        /// <summary>
+        /// Returns a description about the code example.
+        /// </summary>
+        public override string Description
+        {
+            get
+            {
+                return "Shows how to download a set of reports from a list of accounts in " +
+                       "parallel. If you need to obtain a list of accounts, please see the " +
+                       "GetAccountHierarchy or ListAccessibleCustomers examples.";
+            }
+        }
+
+        /// <summary>
+        /// Runs the code example.
+        /// </summary>
+        /// <param name="client">The Google Ads client.</param>
+        /// <param name="customerIds">The Google Ads customer Id.</param>
+        /// <param name="loginCustomerId">Optional login customer ID if your access to the CIDs
+        ///     is via a manager account.</param>
+        public void Run(GoogleAdsClient client, string customerIds, string loginCustomerId = null)
+        {
+            // If a manager ID is supplied, update the login credentials.
+            if (loginCustomerId != null)
+            {
+                client.Config.LoginCustomerId = loginCustomerId;
+            }
+
+            // Get the GoogleAdsService. A single client can be shared by all threads.
+            GoogleAdsServiceClient googleAdsService = client.GetService(
+                Services.V3.GoogleAdsService);
+
+            List<long> ids = customerIds.Split(',').Select(long.Parse).ToList();
+
+            try
+            {
+                // Begin downloading reports and block program termination until complete.
+                Task task = RunDownloadParallelAsync(googleAdsService, ids);
```

Comment:
Can we use Parallel.Foreach instead? It's much cleaner, plus the paralleloptions has a MaxDegreeOfConcurrency that helps with rate limits.

https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.parallel.foreach?view=netcore-3.1

Reply:
https://docs.microsoft.com/en-us/dotnet/standard/parallel-programming/how-to-write-a-simple-parallel-foreach-loop

Reply:
Done.